finite_difference/src/subdomain_mod.f90

```diff
@@ -0,0 +1,302 @@
+!> Module containing the definition of the sub-domain type.
+module subdomain_mod
+  use region_mod
+  implicit none
+
+  !> Type encapsulating the information required to define a single
+  !! sub-domain
+  type :: subdomain_type
+     !> Position of bottom-left and top-right corners of this sub-domain
+     !! in the global domain.
+     integer :: xstart, ystart, xstop, ystop
+     !> Full width and height of this subdomain - includes all halo and
+     !! boundary points. n{x,y} != {x,y}stop - {x,y}start +1 because
+     !! {x,y}start and {x,y}stop are the coordinates of the internal
+     !! region in the global domain.
+     integer :: nx, ny
+     !> The internal region of this subdomain (excluding halo and
+     !! boundary points)
+     type(region_type) :: internal
+  end type subdomain_type
+
+  !> Type encapsulating all information regarding a regular, 2D
+  !! domain decomposition
+  type :: decomposition_type
+     !> Dimensions of the grid of sub-domains
+     integer :: nx, ny
+     !> Number of sub-domains (=nx*ny)
+     integer :: ndomains
+     !> Max dimensions of any sub-domain
+     integer :: max_width, max_height
+     !> Array of the sub-domain definitions
+     type(subdomain_type), allocatable :: subdomains(:)
+  end type decomposition_type
+
+contains
+
+  function decompose(domainx, domainy,             &
+                     ndomains, ndomainx, ndomainy, &
+                     halo_width) result(decomp)
+    !> Decompose a domain consisting of domainx x domainy points
+    !! into a 2D grid.
+    !! Returns an array of tiles describing each of the subdomains.    
```

Comment:
The comment must go before the function, otherwise doxygen does not pick it up.

Reply:
Done.